src/drivers/esp8266_drv.c

```diff
@@ -127,102 +195,157 @@ static const char* get_channel_name(const size_t chan_id)
 
 static void _tx_char_cb(xQueueHandle queue, void *post_tx_arg)
 {
-        struct _channel *ch = post_tx_arg;
-
-        /*
-         * Because we are mimicking a Serial device on top of an actual
-         * socket we need to do some dirty tricks to get the same behavior.
-         * If we have a good connection, then buffer it and send it as
-         * needed.
-         * If we aren't connected we dump all the data on the floor to prevent
-         * the unnecessary stalling of pipelines.  This is on par with
-         * Serial device behavior, but is not ideal for obvious reasons.
-         * We need to move to a socket model before we can fix this properly.
-         * But that will take some work. ITMT this is a good solution until
-         * we get there.
-         */
-        if (ch->connected) {
-                ++ch->tx_chars_buffered;
-        } else {
-                char c;
-                while(pdTRUE == xQueueReceive(queue, &c, 0));
-                ch->tx_chars_buffered = 0;
-        }
+        struct channel *ch = post_tx_arg;
+        ++ch->tx_chars_buffered;
+        cmd_set_check(CHECK_DATA);
 }
 
 static struct Serial* setup_channel_serial(const unsigned int i)
 {
         const char* name = get_channel_name(i);
-        struct _channel* ch =state.channels + i;
-        struct Serial *s = serial_create(name, _SERIAL_TX_BUFF_SIZE,
-                                         _SERIAL_RX_BUFF_SIZE, NULL, NULL,
+        struct channel* ch =state.comm.channels + i;
+        struct Serial *s = serial_create(name, SERIAL_TX_BUFF_SIZE,
+                                         SERIAL_RX_BUFF_SIZE, NULL, NULL,
                                          _tx_char_cb, ch);
         if (s)
                 return s;
 
         /* Fail state */
-        pr_error(_LOG_PFX "Failed to create serial port\r\n");
+        pr_warning(LOG_PFX "Failed to create serial port\r\n");
         return NULL;
 }
 
-static struct _channel* get_channel_for_use(const unsigned int i)
+static struct channel* get_channel_for_use(const unsigned int i)
 {
         if (!is_valid_socket_channel_id(i))
                 return NULL;
 
-        struct _channel* ch =state.channels + i;
-        if (!ch->serial) {
-                ch->serial = setup_channel_serial(i);
-                if (!ch->serial)
-                        return NULL;
-        }
+        struct channel* ch = state.comm.channels + i;
+        if (ch->serial)
+                return ch;
 
-        /* Prep the channel for use */
-        ch->tx_chars_buffered = 0;
-        ch->in_use = true;
-        ch->connected = false; /* Not connected until we say so */
-
-        return ch;
+        /* If here, channel has no serial.  Try to set one up */
+        ch->serial = setup_channel_serial(i);
```

Comment:
could these serial assignments be done in advance when everything is set up upon system init?